### PR TITLE
Add job sidecar terminator controller

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: mingzhou.swx <mingzhou.swx@alibaba-inc.com>

### Ⅰ. Describe what this PR does
People can use this controller to terminate job sidecar containers when other main containers exited.

Just add env `KRUISE_TERMINATE_SIDECAR_WHEN_JOB_EXIT="true"` to sidecar containers to enable this feature.

### Ⅱ. Does this pull request fix one issue?

#1050

### Ⅲ. Describe how to verify it
UT & E2E

### CODE REVIEW

File: pkg/controller/sidecarterminator/virtual_kubelet_handler.go [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+	}
+
+	if isVirtualKubelet(node) {
+		return true, nil
+	}
+	return false, nil
+}
+
+func isVirtualKubelet(node *corev1.Node) bool {
+	if node.Labels[DefaultVKLabelKey] == DefaultVKLabelValue {
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/sidecarterminator/virtual_kubelet_handler.go:46]: ignore

[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/sidecarterminator/virtual_kubelet_handler.go:46]: ignoreall

[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/sidecarterminator/virtual_kubelet_handler.go:46]: ignore

File: config/rbac/role.yaml [hunk 2]
```diff
@@ -25,6 +25,14 @@ rules:
...
   - list
   - watch
+- apiGroups:
+  - ""
+  resources:
+  - pods
+  verbs:
+  - get
+  - update
+  - watch
```
Comment (zmberg) [inline config/rbac/role.yaml:35]: This part is repeated.

File: pkg/controller/sidecarterminator/sidecar_terminator_controller.go [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+
+var (
+	concurrentReconciles = 3
+)
+
+// Sidecar container environment variable definitions which are used to enable SidecarTerminator to take effect on the sidecar container.
+const (
+	// KruiseTerminateSidecar is an env name, which represents a switch to enable sidecar terminator.
+	// The value of this env indicates the specific strategy, could be:
+	// 1. "", "FALSE", "False", "false", which mean this controller will not work on this sidecar.
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_controller.go:54]: Is that only 'True' is considered to be turned on SidecarTerminator and other values are ignored and considered not to be turned on?

File: pkg/controller/sidecarterminator/sidecar_terminator_controller.go [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+var _ reconcile.Reconciler = &ReconcileSidecarTerminator{}
+
+// ReconcileSidecarTerminator reconciles a SidecarTerminator object
+type ReconcileSidecarTerminator struct {
+	client.Client
+	recorder record.EventRecorder
+	scheme   *runtime.Scheme
+}
+
+// +kubebuilder:rbac:groups="",resources=pods,verbs=get;watch;update
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_controller.go:125]: I think you can remove this line.

File: pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+			NamespacedName: types.NamespacedName{
+				Namespace: newPod.Namespace,
+				Name:      newPod.Name,
+			},
+		})
+	}
+}
+
+func isInterestingPod(pod *corev1.Pod) bool {
+	if pod.DeletionTimestamp != nil ||
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go:76]: kubecontroller.IsPodActive(pod)

File: pkg/controller/sidecarterminator/sidecar_terminator_controller.go [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+
+	return r.doReconcile(pod)
+}
+
+func (r *ReconcileSidecarTerminator) doReconcile(pod *corev1.Pod) (reconcile.Result, error) {
+	if !isInterestingPod(pod) {
+		return reconcile.Result{}, nil
+	}
+
+	if allSidecarsHaveBeenCompleted(pod) {
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_controller.go:156]: Can this function be called in isInterestingPod?

File: pkg/controller/sidecarterminator/sidecar_terminator_controller.go [hunk 7]
```diff
@@ -0,0 +1,270 @@
...
+		klog.V(3).Infof("SidecarTerminator -- all sidecars of pod(%v/%v) have been completed, no need to process", pod.Namespace, pod.Name)
+		return reconcile.Result{}, nil
+	}
+
+	if pod.Spec.RestartPolicy == corev1.RestartPolicyOnFailure && !allMainContainersHaveBeenSucceeded(pod) {
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_controller.go:161]: Is it no longer necessary to judge here?

File: pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go (hunk 5 above)
Comment (veophi) [reply pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go:76]: `IsPodActive` is not very exactly to describe this condition.

Comment (veophi) [reply pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go:76]: `IsPodActive` is not very exactly to describe this condition.

File: pkg/controller/sidecarterminator/inplace_update_action.go [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+		klog.V(3).Infof("SidecarTerminator -- InPlace update pod(%v/%v) successfully",
+			originalPod.Namespace, originalPod.Name)
+	}
+
+	return err
+}
+
+// updateSidecarsForInPlaceUpdate do the following things:
+// 1. replace sidecar image with KruiseTerminateSidecarWithImageEnv
+// 2. modify sidecar command and args via modifyCommandViaDownwardAPI if needs
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/inplace_update_action.go:77]: remove the comments.

File: pkg/controller/sidecarterminator/inplace_update_action.go [hunk 9]
```diff
@@ -0,0 +1,113 @@
...
+		pod = &corev1.Pod{}
+		getErr := r.Client.Get(context.TODO(), types.NamespacedName{Name: originalPod.Name, Namespace: originalPod.Namespace}, pod)
+		if getErr != nil {
+			if errors.IsNotFound(getErr) {
+				return nil
+			}
+			return getErr
+		}
+
+		if podHasBeenUpdated(pod, uncompletedSidecars) {
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/inplace_update_action.go:52]: remove the function?

File: pkg/controller/sidecarterminator/kill_container_action.go [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+	err := r.Get(context.TODO(), types.NamespacedName{Namespace: pod.Namespace, Name: getCRRName(pod)}, existingCRR)
+	if err == nil {
+		klog.V(3).Infof("SidecarTerminator -- CRR(%s/%s) exists, waiting for this CRR to complete", existingCRR.Namespace, existingCRR.Name)
+		return nil
+	} else if client.IgnoreNotFound(err) != nil {
+		klog.V(3).Infof("SidecarTerminator -- Error occurred when try to get CRR(%s/%s), error: %v", existingCRR.Namespace, existingCRR.Name, err)
+		return err
+	}
+
+	r.recorder.Eventf(pod, corev1.EventTypeNormal, "SidecarTerminator",
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/kill_container_action.go:49]: move it to 80 line?

File: pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go [hunk 11]
```diff
@@ -0,0 +1,143 @@
...
+	}
+	return containersCompleted(pod, getMain(pod))
+}
+
+func mainContainerStatusChanged(oldPod, newPod *corev1.Pod) bool {
+	sidecars := getSidecar(newPod)
+	for i := range newPod.Status.ContainerStatuses {
+		newStatus := &newPod.Status.ContainerStatuses[i]
+		if !sidecars.Has(newStatus.Name) {
+			oldStatus := getContainerStatus(oldPod.Status.ContainerStatuses, newStatus.Name)
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go:95]: A more stringent judgment that the main container has exited properly?

File: pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go [hunk 12]
```diff
@@ -0,0 +1,143 @@
...
+			if !reflect.DeepEqual(newStatus, oldStatus) {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+func getContainerStatus(Statuses []corev1.ContainerStatus, name string) *corev1.ContainerStatus {
```
Comment (zmberg) [inline pkg/controller/sidecarterminator/sidecar_terminator_pod_event_handler.go:104]: use util function util.GetContainerStatus("app", pod).

File: pkg/controller/sidecarterminator/sidecar_terminator_controller.go [hunk 13]
```diff
@@ -0,0 +1,209 @@
...
+			// Object not found, return.  Created objects are automatically garbage collected.
+			// For additional cleanup logic use finalizers.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	start := time.Now()
+	klog.V(3).Infof("SidecarTerminator -- begin to process pod %v for reconcile", request.NamespacedName)
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/sidecarterminator/sidecar_terminator_controller.go:119]: ignore

**Review by furykerry — COMMENTED** [review]
/approve

**Pull request merged**